### ISSUE TITLE
Different results although rstate set (random seed) 

### ISSUE THREAD

**Issue by Arktius** [hyperopt/hyperopt#640]
Hi folks, 

I have the problem that I get different results with each execution.  I've checked the hyperparameters and they are identical when I repeat it. I've also trained my models without using HPO and got exactly the same loss each time. So...where could the bug be? 

The code looks like this:

```
def ae(p):
    np.random.seed(2020)
    random.seed(2020)
    tf.compat.v1.set_random_seed(2020)
    ## Train model and return loss
     print( rmse, mmse, np.random.get_state()[1][:5])
     
```

```
rstate = np.random.RandomState(2020) 
parameterspace = { 'parameters':{
                'ac':hp.choice('AFunction',['elu','relu','selu','sigmoid','softplus','softsign','tanh']),
                'bn':hp.choice('BatchNorm',[False,True]),
                'bs':hp.quniform('BatchSize', 16, 128,16),
                'bo':hp.quniform('Bottleneck', 8,32,8),
                'da':hp.choice('Dataset',['ml100k']),
                'den':hp.quniform('DecNeurons', 160, 400, 16),
                'dec':hp.loguniform('Decay',np.log(1e-5),np.log(1e-2)),
                'dr':hp.loguniform('Dropout', np.log(1e-6), np.log(0.5)),
                'en':hp.quniform('EncNeurons', 40,128,16),
                'gn':hp.quniform('GaussianNoise',0,0.05,0.0025),
                'lr':hp.loguniform('LR', np.log(1e-3), np.log(0.5)),
                'la':hp.quniform('Layers', 0,4,1),
                'mp':hp.choice('MP', [False,True]),
                'no':hp.choice('Normalize', [False,True]),
                'op':hp.choice('Optimizer', ['adadelta','adagrad','adam','rmsprop','sgd']),
                'sp':hp.choice('Split',['Movie']),
                'we':hp.choice('WeightInit', [0])
                }}

bayes_trials = Trials()
algo = partial(tpe.suggest, n_startup_jobs=3, n_EI_candidates=50) #random points and take best candidate out of choice

nevals = 0
maxevals = 5
failed = 0
while nevals < maxevals:
    best = fmin(fn=ae,space=parameterspace,algo=algo,trials = bayes_trials,max_evals=maxevals+failed,
                rstate = rstate)
    nevals = sum([1 for x in bayes_trials.results if x['status']=='ok'])
    failed = maxevals-nevals

print(best)
```

Results:
[Uploading New Text Document.txt…]()

**Comment by bjkomer** [comment]
My best guess is that since hyperopt is still giving the same sequence of hyperparameters to your function, the bug may be related to how tensorflow handles random seeds. From the docs of [`tf.compat.v1.set_random_seed`](https://www.tensorflow.org/api_docs/python/tf/compat/v1/set_random_seed) it looks like there is both an op level and a graph level seed, but this function only sets the graph one. This makes me suspect that a script with a loop running X configurations will give a different result from a script that evaluates a single configuration run X times (because the effective starting op seed for the run will be different). How are you testing that an evaluation with the same hyperparameters are different? The tf docs aren't clear whether the default op level seed is deterministically chosen or not, this could also affect things.

Another possibility is somehow the seeds used in hyperopt and tensorflow are interacting.

**Comment by Arktius** [comment]
Thanks for your comment. I found the solution or rather a workaround yesterday.  To solve this issue, I did the following steps:

1. Paste these lines before creating your model. (Sets random seeds)
```
np.random.seed(2020)
random.seed(2020)
tf.compat.v1.set_random_seed(2020)
```
2. Paste these lines after training your model. (Resets graph/session)
```
tf.compat.v1.reset_default_graph()
tf.keras.backend.clear_session()
del model,trainx,trainy,valx,valy,testx,testy
gc.collect()
```
After step 1 & 2, one will get identical results when executing the whole script multiple times. There are two exceptions though, namely the optimizers adam and rmsprop. adam and rmsprop have a parameter epsilon which avoids divisions by zero. I've figured out that setting epsilon to ~1.5 reduces the variance within each run. I've trained 2x100 different models and checked the mean and the median accuracy. Both were lower than with the default value. The variance was also much much lower. I think the highest difference between the two runs was something around 1e-6.
